U8Int32[] data